### PR TITLE
Criteria 5 Implementation

### PR DESCRIPTION
**Pull request by lucasmbrown**
Still a WIP, but this is the first draft!

### CODE REVIEW

File: covid/extract.py [hunk 1]
```diff
@@ -172,13 +172,14 @@ def extract_cdc_ili_data():
     )
     filenames_to_contents_map = unzip_string(response.content)
 
-    df = pd.read_csv(BytesIO(filenames_to_contents_map[ILI_NET_CSV]))
-    df = df.set_index("State")
+    df = pd.read_csv(
+        filepath_or_buffer=BytesIO(filenames_to_contents_map[ILI_NET_CSV]), skiprows=1
```
Comment (smacpher) [inline covid/extract.py:176]: Ahh nice `skiprows` is clean!

File: covid/extract.py [hunk 2]
```diff
@@ -187,7 +188,7 @@ def extract_current_cdc_data():
     return cdc_df
 
 
-def extract_historical_cdc_data(credentials):
+def extract_cdc_beds_historical_data(credentials):
```
Comment (smacpher) [inline covid/extract.py:191]: 👍

File: covid/transform.py [hunk 3]
```diff
@@ -187,59 +205,75 @@
...
 
     # Use a multi-index for state and date.
-    df.set_index(keys=[STATE_SOURCE_FIELD, DATE_SOURCE_FIELD], inplace=True)
+    covidtracking_df.set_index(
+        keys=[STATE_SOURCE_FIELD, DATE_SOURCE_FIELD], inplace=True
+    )
 
     # Sort by the index: state ascending, date ascending.
-    df = df.sort_index()
+    covidtracking_df = covidtracking_df.sort_index()
```
Comment (smacpher) [inline covid/transform.py:263]: Nice renaming!

File: covid/transform.py [hunk 4]
```diff
@@ -375,207 +423,338 @@ def transform_covidtracking_data(df):
...
 
     # Create a new multi-index containing the state name (`REGION`) and timestamp of the week.
-    df = df.set_index(keys=[STATE_FIELD, DATE_SOURCE_FIELD])
+    ili_df = ili_df.set_index(keys=[STATE_FIELD, DATE_SOURCE_FIELD])
+
+    # Sort by the index: state ascending, date ascending.
+    ili_df = ili_df.sort_index()
+
+    # Replace `X` with null.
+    ili_df = ili_df.replace(to_replace="X", value=None)
```
Comment (smacpher) [inline covid/transform.py:698]: Ahh yes nice!

File: covid/transform.py [hunk 5]
```diff
@@ -375,207 +423,338 @@ def transform_covidtracking_data(df):
...
+
+        # Calculate change in percent ILI
+        ili_df.loc[(state,), PERCENT_ILI_SPLINE_DIFF] = (
+            ili_df.loc[(state,), PERCENT_ILI_SPLINE].diff(periods=1).values
+        )
+
+        # Calculate criteria 5A: must see two consecutive declines in weekly total ILI data.
+        ili_df.loc[
+            (state,), MAX_RUN_OF_DECREASING_TOTAL_ILI_SPLINE_DIFF
+        ] = get_max_run_in_window(
```
Comment (smacpher) [inline covid/transform.py:731]: These helper functions are so handy 👍

Comment (smacpher) [reply covid/transform.py:731]: What does `get_max_run_in_window` do exactly?

File: covid/transform.py [hunk 6]
```diff
@@ -375,207 +423,338 @@ def transform_covidtracking_data(df):
...
+            series_=ili_df.loc[(state,), PERCENT_ILI_SPLINE_DIFF],
+            positive_values=False,
+            window_size=2,
+        ).values
+
+        ili_df.loc[(state,), CDC_CRITERIA_5C_14_DAY_DECLINE_PERCENT_ILI] = (
+            ili_df.loc[(state,), MAX_RUN_OF_DECREASING_PERCENT_ILI_SPLINE_DIFF] >= 2
+        ).values
+
+        # Calculate criteria 5B: current cases must be lower than cases 2 weeks ago.
```
Comment (smacpher) [inline covid/transform.py]: Nit – should this comment be `5D`?

File: covid/transform.py [hunk 7]
```diff
@@ -94,6 +94,24 @@
...
+TOTAL_ILI = "ILITOTAL"
+TOTAL_ILI_TODAY_MINUS_TOTAL_ILI_14_DAYS_AGO = "Total ILI today minus 2 weeks ago"
+TOTAL_ILI_SPLINE = "Total ILI (WCS)"
+TOTAL_ILI_SPLINE_DIFF = "Change in Total ILI (WCS)"
+MAX_RUN_OF_DECREASING_TOTAL_ILI_SPLINE_DIFF = "Max Run of Decreasing Total ILI (WCS)"
+CDC_CRITERIA_5A_14_DAY_DECLINE_TOTAL_ILI = "CDC Criteria 5A"
+CDC_CRITERIA_5B_OVERALL_DECLINE_TOTAL_ILI = "CDC Criteria 5B"
+CDC_CRITERIA_5C_14_DAY_DECLINE_PERCENT_ILI = "CDC Criteria 5C"
+CDC_CRITERIA_5D_OVERALL_DECLINE_PERCENT_ILI = "CDC Criteria 5D"
+CDC_CRITERIA_5_COMBINED = "CDC Criteria 5 (Partically combined, 5A-5D)"
```
Comment (smacpher) [inline covid/transform.py]: Typo – `partically` should be `partially`, right?

File: main.py [hunk 8]
```diff
@@ -37,25 +41,34 @@ def extract_transform_and_load_covid_data():
...
-    cdc_current_df = extract_current_cdc_data()
-    cdc_historical_df = extract_historical_cdc_data(credentials)
+    cdc_beds_current_df = extract_cdc_beds_current_data()
+    cdc_beds_historical_df = extract_cdc_beds_historical_data(credentials)
     covidtracking_df = extract_covidtracking_historical_data()
+    cdc_ili_df = extract_cdc_ili_data()
 
-    transformed_cdc_df = transform_cdc_data(cdc_current_df, cdc_historical_df)
-    transformed_covidtracking_df = transform_covidtracking_data(df=covidtracking_df)
+    transformed_cdc_beds_df = transform_cdc_beds_data(
```
Comment (smacpher) [inline main.py:48]: Nice, this feels very organized and easy to follow!

**Review by smacpher — COMMENTED** [review]
Done with a first pass! This looks fantastic – very organized and easy to follow. I left a couple of nits and will approve once those are addressed!

File: covid/transform.py (hunk 5 above)
Comment (lucasmbrown) [reply covid/transform.py:731]: It finds the maximum number of either consecutive positive or consecutive negative (depending on the parameter) numbers in a row within a range of recent values!

**Review by ghost — APPROVED** [review]
LGTM!

**Review by smacpher — APPROVED** [review]
LGTM!

**Pull request merged**